RecoEgamma/EgammaElectronAlgos/interface/ElectronHcalHelper.h

```diff
@@ -27,58 +27,51 @@ class ElectronHcalHelper {
-  double hcalESumDepth1(const reco::SuperCluster &, const std::vector<CaloTowerDetId> *excludeTowers = nullptr) const;
-  double hcalESumDepth2(const reco::SuperCluster &, const std::vector<CaloTowerDetId> *excludeTowers = nullptr) const;
-  double hOverEConeSize() const { return cfg_.hOverEConeSize; }
+  void beginEvent(const edm::Event &evt, const edm::EventSetup &eventSetup);
 
-  // Behind clusters
   inline auto hcalTowersBehindClusters(const reco::SuperCluster &sc) const { return egamma::towersOf(sc, *towerMap_); }
-  inline auto hcalESumDepth1BehindClusters(const std::vector<CaloTowerDetId> &towers) const {
-    return egamma::depth1HcalESum(towers, *towersFromCollection_);
-  }
-  inline auto hcalESumDepth2BehindClusters(const std::vector<CaloTowerDetId> &towers) const {
-    return egamma::depth2HcalESum(towers, *towersFromCollection_);
-  }
+  double hcalESum(const reco::SuperCluster &, int depth) const;
+  double hOverEConeSize() const { return cfg_.hOverEConeSize; }
+  int maxSeverityHB() const { return cfg_.maxSeverityHB; }
+  int maxSeverityHE() const { return cfg_.maxSeverityHE; }
 
   // forward EgammaHadTower methods, if checkHcalStatus is enabled, using towers and H/E
-  // otherwise, return true
   bool hasActiveHcal(const reco::SuperCluster &sc) const;
 
+  // QoL when one needs raw instances of EgammaHcalIsolation in addition to this class
+  const auto hcalTopology() const { return hcalTopology_; }
+  const auto hcalChannelQuality() const { return hcalChannelQuality_; }
+  const auto hcalSevLvlComputer() const { return hcalSevLvlComputer_; }
+  const auto towerMap() const { return towerMap_; }
+
 private:
   const Configuration cfg_;
 
-  edm::ESGetToken<HcalChannelQuality, HcalChannelQualityRcd> hcalChannelQualityToken_;
+  edm::ESGetToken<CaloGeometry, CaloGeometryRecord> caloGeometryToken_;
   edm::ESGetToken<HcalTopology, HcalRecNumberingRecord> hcalTopologyToken_;
+  edm::ESGetToken<HcalChannelQuality, HcalChannelQualityRcd> hcalChannelQualityToken_;
+  edm::ESGetToken<HcalSeverityLevelComputer, HcalSeverityLevelComputerRcd> hcalSevLvlComputerToken_;
   edm::ESGetToken<CaloTowerConstituentsMap, CaloGeometryRecord> towerMapToken_;
-  edm::ESGetToken<CaloGeometry, CaloGeometryRecord> caloGeometryToken_;
 
   // event data (rechits strategy)
-  std::unique_ptr<EgammaHcalIsolation> hcalIso_ = nullptr;
-
-  // event data (towers strategy)
-  std::unique_ptr<EgammaTowerIsolation> towerIso1_ = nullptr;
-  std::unique_ptr<EgammaTowerIsolation> towerIso2_ = nullptr;
-  CaloTowerCollection const *towersFromCollection_ = nullptr;
-  CaloTowerConstituentsMap const *towerMap_ = nullptr;
-  HcalChannelQuality const *hcalQuality_ = nullptr;
+  std::unique_ptr<EgammaHcalIsolation> hcalIso_;
   HcalTopology const *hcalTopology_ = nullptr;
+  HcalChannelQuality const *hcalChannelQuality_ = nullptr;
+  HcalSeverityLevelComputer const *hcalSevLvlComputer_ = nullptr;
+  CaloTowerConstituentsMap const *towerMap_ = nullptr;
```

Comment:
I would like to understand why you still need from the CaloTowerConstituentsMap.
Here and in all other files.

Reply:
For matching the ecal and hcal rechits behind cluster. See ```EgammaHcalIsolation::getHcalESumBc()```

Reply:
why this is done twice  in the ElectronHcalHelper and the EgammaHcalIsolation.h

I see that for photon and GSFElectron you have `hcalTowersBehindClusters()`
https://github.com/cms-sw/cmssw/pull/33666/files#diff-2a8819f54a944ade88b9fc2dd659ed4b5c16725c31a250abd7ef89860e908dcdR459

and for Photons you found a way to avoid use the CaloTowerConstituentsMap
https://github.com/cms-sw/cmssw/pull/33666/files#diff-fd3b0e20a0dc203a5747ecc1765f08c85018edee832ab5967e260d1f7d06bb35L353

Reply:
ElectronHcalHelper uses EgammaHcalIsolation under the hood.

And PhotonProducer doesn't explicitly show CaloTowerConstituentsMap usage only because ElectronHcalHelper takes care of this internally, while EgammaHcalIsolation does not.

Reply:
is this still an issue? I could ask for a bit more context to understand the problem.